src/components/Project/ProjectBasics/ProjectForm/ProjectForm.tsx

```diff
@@ -54,6 +56,116 @@ const ProjectForm = () => {
+        data = {...data, "finances": updatedFinances}
+        finances = updatedFinances;
+      }
+    }
+    if (data.estPlanningEnd) {
+      const previousPlanningEndYear = getYear(project.estPlanningEnd);
+      const planningEndYear = getYear(data.estPlanningEnd);
+      const constructionStartYear = getYear(project.estConstructionStart);
+      // If new planning end year is smaller that the previous one, budget from the years that are not within the schedule need to
+      // be moved backwards to the new end year
+      if (previousPlanningEndYear && planningEndYear && planningEndYear < previousPlanningEndYear) {
+        // If there was an overlap between planning end year and construction start year, budget shouldn't be moved from that year
+        // but still needs to be moved from all the years that are not within the schedule anymore
+        if (constructionStartYear == previousPlanningEndYear &&  previousPlanningEndYear - planningEndYear > 1) {
+          const updatedFinances = moveBudgetBackwards(finances, previousPlanningEndYear - 1, planningEndYear);
+          finances = updatedFinances;
+        } else if (constructionStartYear != previousPlanningEndYear) {
+          const updatedFinances = moveBudgetBackwards(finances, previousPlanningEndYear, planningEndYear);
+          data = {...data, "finances": updatedFinances};
+          finances = updatedFinances;
+        }
+      }
+    }
+    if (data.estConstructionStart) {
+      const previousConstructionStartYear = getYear(project.estConstructionStart);
+      const constructionStartYear = getYear(data.estConstructionStart);
+      const planningEndYear = getYear(project.estPlanningEnd);
+      // If new construction start year is bigger than the previous one, budget from years that are not within the schedule of the project
+      // need to be moved to the new construction start year
+      if (previousConstructionStartYear && constructionStartYear && constructionStartYear > previousConstructionStartYear) {
+        // If there was an overlap between planning end year and construction start year, budget shouldn't be moved from that year
+        // but still needs to be moved from all the years that are not within the schedule anymore
+        if (planningEndYear == previousConstructionStartYear && constructionStartYear - previousConstructionStartYear > 1) {
+          const updatedFinances = moveBudgetForwards(finances, previousConstructionStartYear + 1, constructionStartYear);
+          data = {...data, "finances": updatedFinances};
+          finances = updatedFinances;
+        } else if (planningEndYear != previousConstructionStartYear) {
+          const updatedFinances = moveBudgetForwards(finances, previousConstructionStartYear, constructionStartYear);
+          data = {...data, "finances": updatedFinances};
+          finances = updatedFinances;
+        }
+      }
```

Comment:
This could be refactored maybe a bit e.g. something like
```
        let updatedFinances;
        const endYearIsSomething = planningEndYear == previousConstructionStartYear && constructionStartYear - previousConstructionStartYear > 1;
        const planningYearNotMatchingPreviousConstructionStartYear = planningEndYear != previousConstructionStartYear;
        
        if (endYearIsSomething) {
          updatedFinances = moveBudgetForwards(finances, previousConstructionStartYear + 1, constructionStartYear);
        } else if (planningYearNotMatchingPreviousConstructionStartYear) {
          updatedFinances = moveBudgetForwards(finances, previousConstructionStartYear, constructionStartYear);
        }

        data = {...data, "finances": updatedFinances};
        finances = updatedFinances;
```

Reply:
I did some refactoring using this as a reference. Do you think it's better now or would you still change something?

Reply:
It looks good now! :smile: